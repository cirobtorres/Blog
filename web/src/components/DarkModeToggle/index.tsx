"use client";

import { IoMoon, IoSunny } from "react-icons/io5";
import { useTheme } from "next-themes";
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
} from "../shadcnui/tooltip";

export default function DarkModeToggle() {
  const { setTheme } = useTheme();

  return (
    <TooltipProvider>
      <Tooltip>
        <TooltipTrigger asChild>
          <div className="relative">
            <div className="flex shrink-0 justify-center items-center rounded-full size-8 duration-700 cursor-pointer bg-blog-background-1">
              <button
                type="button"
                onClick={() => setTheme("light")}
                className="absolute transition-all duration-[600ms] opacity-0 scale-0 dark:opacity-100 dark:scale-100"
              >
                <IoMoon className="text-xl text-[#40b88c]" />
              </button>
              <button
                type="button"
                onClick={() => setTheme("dark")}
                className="absolute transition-all duration-[600ms] opacity-100 scale-100 dark:opacity-0 dark:scale-0
                "
              >
                <IoSunny className="text-xl text-[#fbbf24]" />
              </button>
            </div>
          </div>
        </TooltipTrigger>
        <TooltipContent>
          <p>Alternar tema</p>
        </TooltipContent>
      </Tooltip>
    </TooltipProvider>
  );
}
