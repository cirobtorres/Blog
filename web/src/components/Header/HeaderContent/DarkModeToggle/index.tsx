"use client";

import { useTheme } from "next-themes";
import { FaMoon } from "react-icons/fa6";
import { IoSunny } from "react-icons/io5";
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
} from "../../../Shadcnui/tooltip";
import { useEffect, useState } from "react";
import { Skeleton } from "../../../Shadcnui/skeleton";

export default function DarkModeToggle() {
  const { theme, setTheme } = useTheme();
  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    setMounted(true);
  }, []);

  return (
    <>
      {!mounted && (
        <Skeleton
          aria-label="Carregando botÃ£o de modo noturno"
          className="size-8 rounded-full"
        />
      )}
      {mounted && (
        <TooltipProvider>
          <Tooltip>
            <TooltipTrigger
              asChild
              className="outline-blog-foreground-highlight focus:outline-blog-foreground-highlight"
            >
              <button
                data-testid="dark-mode-toggle"
                aria-label="Alternar entre tema claro e escuro"
                role="button"
                className="relative"
                tabIndex={0}
                onClick={(event) => {
                  event?.preventDefault();
                  setTheme(theme === "dark" ? "light" : "dark");
                }}
              >
                <div className="flex shrink-0 justify-center items-center rounded-full size-8 cursor-pointer bg-blog-background-1 dark:bg-[hsl(0,0%,14.9%,0.75)]">
                  <div
                    data-testid="moon-button"
                    tabIndex={-1}
                    className="absolute duration-500 transition-all pointer-events-none p-1"
                    style={{
                      opacity: theme === "dark" ? 1 : 0,
                      scale: theme === "dark" ? 1 : 0,
                      transform:
                        theme === "dark" ? "rotate(360deg)" : "rotate(0deg)",
                    }}
                  >
                    <FaMoon className="text-xl text-[#40b88c]" />
                  </div>
                  <div
                    data-testid="sun-button"
                    tabIndex={-1}
                    className="absolute duration-500 transition-all pointer-events-none p-1"
                    style={{
                      opacity: theme !== "dark" ? 1 : 0,
                      scale: theme !== "dark" ? 1 : 0,
                      transform:
                        theme !== "dark" ? "rotate(0deg)" : "rotate(360deg)",
                    }}
                  >
                    <IoSunny className="text-xl text-[#fbbf24]" />
                  </div>
                  <span
                    data-testid="dark-mode-toggle-span"
                    role="status"
                    aria-live="polite"
                    className="sr-only"
                  >
                    Tema {theme === "dark" ? "escuro" : "claro"}
                  </span>
                </div>
              </button>
            </TooltipTrigger>
            <TooltipContent role="tooltip">
              <p data-testid="dark-mode-toggle-tooltip-text">Alternar tema</p>
            </TooltipContent>
          </Tooltip>
        </TooltipProvider>
      )}
    </>
  );
}
