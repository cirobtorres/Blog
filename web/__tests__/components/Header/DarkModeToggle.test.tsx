import { render, screen, fireEvent } from "@testing-library/react";
import * as nextThemes from "next-themes";
import DarkModeToggle from "../../../src/components/Header/HeaderContent/DarkModeToggle";
import userEvent from "@testing-library/user-event";

jest.mock("next-themes", () => ({
  useTheme: jest.fn(),
}));

function setTheme(theme = "light") {
  const mockSetTheme = jest.fn();

  jest.spyOn(nextThemes, "useTheme").mockImplementation(() => ({
    theme,
    setTheme: mockSetTheme,
    themes: ["light", "dark"],
  }));
  return mockSetTheme;
}

describe("DarkModeToggle", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("renders dark mode toggle button", () => {
    setTheme();
    render(<DarkModeToggle />);
    const toogleButton = screen.getByTestId("dark-mode-toggle");
    expect(toogleButton).toBeInTheDocument();
  });

  it("switches to dark theme if moon button is clicked", () => {
    const mockSetTheme = setTheme();
    render(<DarkModeToggle />);
    const toogleButton = screen.getByTestId("dark-mode-toggle");
    fireEvent.click(toogleButton);
    expect(mockSetTheme).toHaveBeenCalledWith("dark");
  });

  it("switches to dark theme if the sun button is clicked", () => {
    const mockSetTheme = setTheme("dark");
    render(<DarkModeToggle />);
    const toogleButton = screen.getByTestId("dark-mode-toggle");
    fireEvent.click(toogleButton);
    expect(mockSetTheme).toHaveBeenCalledWith("light");
  });

  it("checks the accessibility tools are in place", () => {
    setTheme("dark");
    const { getByTestId } = render(<DarkModeToggle />);

    const toogleButtonDark = getByTestId("dark-mode-toggle");
    const toogleButtonSpanDark = getByTestId("dark-mode-toggle-span");
    const moonButtonDark = getByTestId("moon-button");
    const sunButtonDark = getByTestId("sun-button");

    expect(toogleButtonSpanDark).toHaveClass("sr-only");
    expect(toogleButtonSpanDark).toHaveTextContent("Tema escuro");
    expect(toogleButtonDark).toHaveAttribute(
      "aria-label",
      "Alternar entre tema claro e escuro"
    );
    expect(moonButtonDark).toHaveStyle("opacity: 1");
    expect(moonButtonDark).toHaveStyle("transform: rotate(360deg)");
    expect(sunButtonDark).toHaveStyle("opacity: 0");
    expect(sunButtonDark).toHaveStyle("transform: rotate(360deg)");

    setTheme("light");

    fireEvent.click(toogleButtonDark);
    const { getAllByTestId } = render(<DarkModeToggle />);

    const toogleButtonSpanLight = getAllByTestId("dark-mode-toggle-span")[1];
    const moonButtonLight = getAllByTestId("moon-button")[1];
    const sunButtonLight = getAllByTestId("sun-button")[1];

    expect(toogleButtonSpanLight).toHaveTextContent("Tema claro");
    expect(moonButtonLight).toHaveStyle("opacity: 0");
    expect(moonButtonLight).toHaveStyle("transform: rotate(0deg)");
    expect(sunButtonLight).toHaveStyle("opacity: 1");
    expect(sunButtonLight).toHaveStyle("transform: rotate(0deg)");
  });

  it("should correctly announce the current theme for screen readers", () => {
    setTheme("dark");
    render(<DarkModeToggle />);
    const themeStatus = screen.getByTestId("dark-mode-toggle-span");
    expect(themeStatus).toHaveTextContent("Tema escuro");
    expect(themeStatus).toHaveAttribute("role", "status");
    expect(themeStatus).toHaveAttribute("aria-live", "polite");
    expect(themeStatus).toHaveClass("sr-only");
  });

  it("should update the theme announcement when theme changes", async () => {
    const mockSetTheme = setTheme();
    render(<DarkModeToggle />);
    const themeStatus = screen.getByTestId("dark-mode-toggle-span");
    expect(themeStatus).toHaveTextContent("Tema claro");
    await userEvent.click(screen.getByTestId("dark-mode-toggle"));
    expect(mockSetTheme).toHaveBeenCalledWith("dark");
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(<DarkModeToggle />);
    expect(asFragment()).toMatchSnapshot();
  });
});
