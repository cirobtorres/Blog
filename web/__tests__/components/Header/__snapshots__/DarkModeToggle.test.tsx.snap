// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DarkModeToggle matches the snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="BotÃ£o de alternar tema de claro para escuro"
    class="relative focus:outline-blog-foreground-highlight"
    data-state="closed"
    data-testid="dark-mode-toggle"
    role="group"
    tabindex="0"
  >
    <div
      class="flex shrink-0 justify-center items-center rounded-full size-8 cursor-pointer bg-blog-background-1 dark:bg-[hsl(0,0%,14.9%,0.75)]"
    >
      <div
        class="absolute duration-[600ms] transition-all pointer-events-none p-1"
        data-testid="moon-button"
        style="opacity: 0; transform: rotate(0deg);"
        tabindex="-1"
      >
        <svg
          class="text-xl text-[#40b88c]"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 384 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M223.5 32C100 32 0 132.3 0 256S100 480 223.5 480c60.6 0 115.5-24.2 155.8-63.4c5-4.9 6.3-12.5 3.1-18.7s-10.1-9.7-17-8.5c-9.8 1.7-19.8 2.6-30.1 2.6c-96.9 0-175.5-78.8-175.5-176c0-65.8 36-123.1 89.3-153.3c6.1-3.5 9.2-10.5 7.7-17.3s-7.3-11.9-14.3-12.5c-6.3-.5-12.6-.8-19-.8z"
          />
        </svg>
      </div>
      <div
        class="absolute duration-[600ms] transition-all pointer-events-none p-1"
        data-testid="sun-button"
        style="opacity: 1; transform: rotate(0deg);"
        tabindex="-1"
      >
        <svg
          class="text-xl text-[#fbbf24]"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 118a22 22 0 0 1-22-22V48a22 22 0 0 1 44 0v48a22 22 0 0 1-22 22zm0 368a22 22 0 0 1-22-22v-48a22 22 0 0 1 44 0v48a22 22 0 0 1-22 22zm113.14-321.14a22 22 0 0 1-15.56-37.55l33.94-33.94a22 22 0 0 1 31.11 31.11l-33.94 33.94a21.93 21.93 0 0 1-15.55 6.44zM108.92 425.08a22 22 0 0 1-15.55-37.56l33.94-33.94a22 22 0 1 1 31.11 31.11l-33.94 33.94a21.94 21.94 0 0 1-15.56 6.45zM464 278h-48a22 22 0 0 1 0-44h48a22 22 0 0 1 0 44zm-368 0H48a22 22 0 0 1 0-44h48a22 22 0 0 1 0 44zm307.08 147.08a21.94 21.94 0 0 1-15.56-6.45l-33.94-33.94a22 22 0 0 1 31.11-31.11l33.94 33.94a22 22 0 0 1-15.55 37.56zM142.86 164.86a21.89 21.89 0 0 1-15.55-6.44l-33.94-33.94a22 22 0 0 1 31.11-31.11l33.94 33.94a22 22 0 0 1-15.56 37.55zM256 358a102 102 0 1 1 102-102 102.12 102.12 0 0 1-102 102z"
          />
        </svg>
      </div>
      <span
        aria-live="polite"
        class="sr-only"
        data-testid="dark-mode-toggle-span"
        role="status"
      >
        Tema claro
      </span>
    </div>
  </button>
</DocumentFragment>
`;
